ar-setup.sh

```diff
@@ -0,0 +1,138 @@
+	gcloud beta artifacts repositories create "${app_name}" --location="${artifact_registry_location}" --repository-format=docker >/dev/null 2>&1
+	if [ $? -ne 0 ]; then
+		echo "Failed to create the artifact repository \""${app_name}"\" at the location "${artifact_registry_location}""
+		exit 1
+	else 
+		echo "Created the artifact repository \""${app_name}"\" at the location "${artifact_registry_location}""
+	fi
+else
+	echo "The artifact repository \""${app_name}"\" already exists"
+fi
+
+# check whether the service account already exists, and create the service account if it does not exist
+service_account_name="${app_name}-push"
+service_account_email="${service_account_name}@${project}.iam.gserviceaccount.com"
+gcloud iam service-accounts describe "${service_account_email}" >/dev/null 2>&1
+if [ $? -ne 0 ]; then
+	echo "The service account \""${service_account_email}"\" does not exist"
+	service_account_description="Service Account for accessing the artifact repository \""${app_name}"\"" 
+	gcloud iam service-accounts create "${service_account_name}" --description "${service_account_description}" >/dev/null 2>&1
+	if [ $? -ne 0 ]; then
+		echo "Failed to create the service account \""${service_account_email}"\""
+		exit 1
+	else
+		echo "Created the service account \""${service_account_email}"\""
+	fi
+else
+	echo "The service account \""${service_account_email}"\" already exists"
+fi
+
+# grant the Google service account roles/artifactregistry.writer on the AR repo (this command is idempotent)
+gcloud beta artifacts repositories add-iam-policy-binding "${app_name}" \
+	--location="${artifact_registry_location}" \
+	--member=serviceAccount:"${service_account_email}" \
+	--role=roles/artifactregistry.writer >/dev/null
+if [ $? -ne 0 ]; then
+	echo "Failed to grant the service account "${service_account_email}" \"roles/artifactregistry.writer\" on the artifact repository \""${app_name}"\""
+	exit 1
+else
+	echo "Granted the service account "${service_account_email}" \"roles/artifactregistry.writer\" on the artifact repository \""${app_name}"\""
+fi
+
+add_gitlab_project_vars() {
+	app_config_repo=$1
+
+	# get the gitlab hostname and gitlab project name from app_config_repo
+	https_prefix="https://"
+	https_prefix_len=${#https_prefix}
+	app_config_repo_without_https_prefix=${app_config_repo:https_prefix_len}
+
+	end_of_gitlab_hostname_index=$(expr index ${app_config_repo_without_https_prefix} /)
+	gitlab_hostname_len=https_prefix_len+end_of_gitlab_hostname_index
+	gitlab_hostname=${app_config_repo:0:gitlab_hostname_len}
+
+	gitlab_project=${app_config_repo:gitlab_hostname_len}
+	# replace the slash in the project name with %2F
+	gitlab_project=${gitlab_project//\//%2F}
+
+	# add the AR repo and the service account key into the app project as GitLab project variables
+	output_repo=$(curl -s --header "PRIVATE-TOKEN: ${gitlab_access_token}" "${gitlab_hostname}api/v4/projects/${gitlab_project}/variables/GCP_AR_REPO")
```

Comment:
Make it a readonly variable since it is being used multiple times?

Reply:
Done. Thanks!

Reply:
Sorry, github seems to not highlight my selected words.
I meant making `GCP_AR_REPO` and `GCP_AR_KEY` readonly variables, not those local variables such as `gitlab_hostname` or `output_repo`.

`GCP_AR_REPO` and `GCP_AR_KEY` should be global exposed variables.
Moving them at the beginning of the script instead of embedding inside the method helps to increase the visibility.

Reply:
This could be merged after this one is resolved.

Reply:
> `GCP_AR_REPO` and `GCP_AR_KEY` should be global exposed variables.
> Moving them at the beginning of the script instead of embedding inside the method helps to increase the visibility.

Sure. I can definitely do that. 

> I meant making GCP_AR_REPO and GCP_AR_KEY readonly variables, not those local variables such as gitlab_hostname or output_repo.

I kinda like set these local varaibles to be `readonly`. Is there any side effect for setting them as readonly?

Reply:
@nan-yu, PTAL.